kettle/stream.py

```diff
@@ -0,0 +1,306 @@
+                if not db.insert_build(build_dir, started, finished):
+                    print('already present??')
+                start = time.localtime(started.get('timestamp', 0) if started else 0)
+                print(build_dir, bool(started), bool(finished),
+                      time.strftime('%F %T %Z', start),
+                      finished and finished.get('result'))
+                build_dirs.append(build_dir)
+                acks.append(ack_id)
+            else:
+                print('finished.json missing?', build_dir, started, finished)
+    finally:
+        pool.close()
+    db.commit()
+    return acks, build_dirs
+
+
+def row_to_mapping(row, schema):
+    """Convert a dictionary to a list for bigquery.Table.insert_data.
+
+    Silly. See https://github.com/GoogleCloudPlatform/google-cloud-python/issues/3396
+    """
+    return [row.get(field.name, [] if field.mode == 'REPEATED' else None) for field in schema]
+
+
+def retry(func, *args, **kwargs):
+    """Run a function with arguments, retrying on server errors. """
+    # pylint: disable=no-member
+    for attempt in xrange(20):
+        try:
+            return func(*args, **kwargs)
+        except (socket.error, google.cloud.exceptions.ServerError):
+            # retry with exponential backoff
+            traceback.print_exc()
+            time.sleep(1.4 ** attempt)
+    return func(*args, **kwargs)  # one last attempt
+
+
+def insert_data(table, rows_iter):
+    """Upload rows from rows_iter into bigquery table table.
+
+    rows_iter should return a series of (row_id, row dictionary) tuples.
+    The row dictionary must match the table's schema.
+
+    Returns the row_ids that were inserted.
+    """
+    emitted = set()
+
+    rows = []
+    row_ids = []
+
+    for row_id, row in rows_iter:
+        emitted.add(row_id)
+        if len(json.dumps(row)) > 1e6:
+            print('ERROR: row too long', row['path'])
+            continue
+        row = row_to_mapping(row, table.schema)
+        rows.append(row)
+        row_ids.append(row_id)
+
+    if not rows:  # nothing to do
+        return []
+
+    def insert(table, rows, row_ids):
+        """Insert rows with row_ids into table, retrying as necessary."""
+        errors = retry(table.insert_data, rows, row_ids, skip_invalid_rows=True)
+
+        if not errors:
+            print('Loaded {} builds into {}'.format(len(rows), table.name))
+        else:
+            print('Errors:')
+            pprint.pprint(errors)
+            pprint.pprint(table.schema)
+
+    if len(json.dumps(rows)) > 10e6:
+        print('WARNING: too big for one insert, doing stupid slow version')
+        for row, row_id in zip(rows, row_ids):
+            insert(table, [row], [row_id])
+    else:
+        insert(table, rows, row_ids)
+
+    return emitted
+
+
+def main(db, sub, tables, client_class=make_db.GCSClient, stop=None):
+    # pylint: disable=too-many-locals
+    gcs_client = client_class('', {})
+
+    if not stop:
```

Comment:
I would be explicit here: `if stop is not None:`. Because this is something that returns False iono it might confusingly evaluate to False and make this fire when it shouldn't.